db/background_mgr_resync_dcp.go

```diff
@@ -0,0 +1,282 @@
+	db := options["database"].(*Database)
+	regenerateSequences := options["regenerateSequences"].(bool)
+
+	resyncLoggingID := "Resync: " + r.ResyncID
+
+	persistClusterStatus := func() {
+		err := persistClusterStatusCallback()
+		r.SetStats(int(r.DocsProcessed.Value()), int(r.DocsChanged.Value()))
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to persist cluster status on-demand for resync operation: %v", resyncLoggingID, err)
+		}
+	}
+	defer persistClusterStatus()
+
+	defer atomic.CompareAndSwapUint32(&db.State, DBResyncing, DBOffline)
+
+	var unusedSequences []uint64
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		var err error
+		docID := string(event.Key)
+		key := realDocID(docID)
+		base.TracefCtx(ctx, base.KeyAll, "[%s] Received DCP event %d for doc %v", resyncLoggingID, event.Opcode, base.UD(docID))
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		r.DocsProcessed.Add(1)
+		_, unusedSequences, err = db.updateDocument(ctx, docID, key, regenerateSequences, unusedSequences)
+
+		if err == nil {
+			r.DocsChanged.Add(1)
+		} else if err != base.ErrUpdateCancel {
+			base.WarnfCtx(ctx, "[%s] Error updating doc %q: %v", resyncLoggingID, base.UD(docID), err)
+		}
+		return true
+	}
+
+	collection, err := base.AsCollection(db.Bucket)
+	if err != nil {
+		return err
+	}
+
+	clientOptions, err := getReSyncDCPClientOptions(collection, db.Options.GroupID)
+	if err != nil {
+		return err
+	}
+	clientOptions.InitialMetadata = base.BuildDCPMetadataSliceFromVBUUIDs(r.VBUUIDs)
+
+	dcpFeedKey := generateResyncDCPStreamName(r.ResyncID)
+	base.InfofCtx(ctx, base.KeyAll, "[%s] Starting DCP feed %q for resync", resyncLoggingID, dcpFeedKey)
+	dcpClient, err := base.NewDCPClient(dcpFeedKey, callback, *clientOptions, collection)
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to create resync DCP client! %v", resyncLoggingID, err)
+		return err
+	}
+
+	doneChan, err := dcpClient.Start()
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to start resync DCP feed! %v", resyncLoggingID, err)
+		_ = dcpClient.Close()
+		return err
+	}
+	base.DebugfCtx(ctx, base.KeyAll, "[%s] DCP client started.", resyncLoggingID)
+
+	select {
+	case <-doneChan:
+		base.InfofCtx(ctx, base.KeyAll, "[%s] Finished re-running sync function. %d/%d docs changed", resyncLoggingID, r.DocsChanged.Value(), r.DocsProcessed.Value())
+		err = dcpClient.Close()
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to close resync DCP client! %v", resyncLoggingID, err)
+			return err
+		}
+
+		db.releaseSequences(ctx, unusedSequences)
+		if regenerateSequences {
+			if err := db.updateAllPrincipalsSequences(ctx); err != nil {
```

Comment:
The user/role documents we want to update are part of the DCP feed (we're omitting them on line 124).  Would be preferable to process as part of DCP feed instead of as a second step.

Reply:
I was thinking of processing user/role from DCP feed too but If I understood correctly, current implementation only processes user/role after it completes processing all docs. If the process was interrupted in the middle, it will not release sequences, update user/role and invalidate user cache as per https://github.com/couchbase/sync_gateway/blob/6a13939608760c221fbd436ee595a6b56f3a9b7e/db/database.go#L1747